File: src/core/safety/settingsHierarchy.ts
/** 3-tier settings hierarchy per spec §4.6. */

import { readFile } from "node:fs/promises";
import {
	type AnyVersionedConfig,
	loadConfigWithMigration,
} from "../config/versioning.js";
import { atomicWriteJson } from "../persistence/atomicWrite.js";

export interface SettingsTier {
	source: "user_global" | "project_shared" | "project_local";
	patterns: string[];
	deny: string[];
}

/**
 * Versioned shape of `settings.json` on disk. Currently identical to the
 * config versioning chain target (v2). When new fields are added, bump the
 * versioning chain in `core/config/versioning.ts` and add a migration there;
 * `loadSettings` will pick them up automatically via `loadConfigWithMigration`.
 */
export type Settings = AnyVersionedConfig;

export interface MergedSettings {
	allow: string[];
	deny: string[];
}

export interface MergeOpts {
	/**
	 * Whether the user has explicitly trusted this project. When `false`
	 * (the default — safe for an unknown / freshly-cloned repo),
	 * `project_shared` and `project_local` patterns are IGNORED in the
	 * `allow` set: a hostile repo cannot escalate permissions just by
	 * shipping its own `.claude/settings.json`. All tiers still
	 * contribute to `deny` (any tier can lock something down).
	 *
	 * Callers wire this from the runtime trust check (e.g. presence of
	 * the workspace path in `~/.apohara/trusted-projects`).
	 */
	trustProject?: boolean;
}

export function mergeSettingsTiers(
	tiers: SettingsTier[],
	opts: MergeOpts = {},
): MergedSettings {
	const trustProject = opts.trustProject ?? false;
	const allow = new Set<string>();
	const deny = new Set<string>();
	for (const t of tiers) {
		const contributesAllow = t.source === "user_global" || trustProject;
		if (contributesAllow) {
			for (const p of t.patterns) allow.add(p);
		}
		for (const p of t.deny) deny.add(p);
	}
	return { allow: Array.from(allow), deny: Array.from(deny) };
}

/**
 * Load and migrate `settings.json` from disk.
 *
 * Multica #17 fix (T4.4c): cables the generic versioning chain
 * (`loadConfigWithMigration`, T4.8b) into the settings loader so adding
 * a new field in a release does NOT break existing installs.
 *
 * Behavior:
 *   - Legacy file (no `schema_version`) is auto-promoted to v1 in-place,
 *     then migrated v1 → v2 by the versioning chain.
 *   - Already-versioned file is passed through (and migrated forward if
 *     `schema_version` is older than the chain target).
 *   - A `.bak` snapshot of the pre-migration form is written by
 *     `loadConfigWithMigration` whenever a migration step runs (it also
 *     restores `.bak` on write failure).
 */
export async function loadAndMigrateSettings(path: string): Promise<Settings> {
	const raw = JSON.parse(await readFile(path, "utf-8"));
	// Auto-promote legacy (no schema_version) to v1 so the generic
	// migration chain can take over from a known baseline.
	if (raw.schema_version === undefined) {
		raw.schema_version = 1;
		await atomicWriteJson(path, raw);
	}
	const migrated = await loadConfigWithMigration(path, /*targetVersion=*/ 2);
	return migrated;
}

File: crates/apohara-safety/src/settings_hierarchy.rs
//! 3-tier settings hierarchy — ports `src/core/safety/settingsHierarchy.ts`.
//!
//! NOTE: the TS module also pulls the `loadConfigWithMigration` chain
//! from `core/config/versioning.ts`. That chain belongs to a different
//! crate (config versioning is workspace-wide, not safety-specific) and
//! will land separately; this port covers the merge logic, which is
//! what the permission service consumes.

use serde::{Deserialize, Serialize};
use std::collections::BTreeSet;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SettingsSource {
    UserGlobal,
    ProjectShared,
    ProjectLocal,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SettingsTier {
    pub source: SettingsSource,
    pub patterns: Vec<String>,
    pub deny: Vec<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
pub struct MergedSettings {
    pub allow: Vec<String>,
    pub deny: Vec<String>,
}

#[derive(Debug, Clone, Copy, Default)]
pub struct MergeOpts {
    /// Whether the user has explicitly trusted this project. Default
    /// `false` so a hostile project cannot escalate permissions just by
    /// shipping its own `.claude/settings.json`. All tiers still
    /// contribute to `deny`.
    pub trust_project: bool,
}

pub fn merge_settings_tiers(tiers: &[SettingsTier], opts: MergeOpts) -> MergedSettings {
    // BTreeSet to keep the output deterministic across runs — TS's
    // `Set` iteration order is insertion order, but BTreeSet gives us
    // sort order which matters for snapshot tests and audit diffs.
    let mut allow: BTreeSet<String> = BTreeSet::new();
    let mut deny: BTreeSet<String> = BTreeSet::new();
    for t in tiers {
        let contributes_allow = matches!(t.source, SettingsSource::UserGlobal) || opts.trust_project;
        if contributes_allow {
            for p in &t.patterns {
                allow.insert(p.clone());
            }
        }
        for p in &t.deny {
            deny.insert(p.clone());
        }
    }
    MergedSettings {
        allow: allow.into_iter().collect(),
        deny: deny.into_iter().collect(),
    }
}
